Index: src/libnet_build_ip.c
===================================================================
--- src/libnet_build_ip.c	(revision 382)
+++ src/libnet_build_ip.c	(working copy)
@@ -557,8 +557,12 @@
     }
 
     /* no checksum for IPv6 */
-    return (ptag ? ptag : libnet_pblock_update(l, p, LIBNET_IPV6_H,
-            LIBNET_PBLOCK_IPV6_H));
+    ptag = ptag ? ptag : libnet_pblock_update(l, p, LIBNET_IPV6_H,
+            LIBNET_PBLOCK_IPV6_H);
+
+    libnet_pblock_record_ip_offset(l, p);
+
+    return ptag;
 bad:
     libnet_pblock_delete(l, p);
     return (-1);
Index: src/libnet_pblock.c
===================================================================
--- src/libnet_pblock.c	(revision 382)
+++ src/libnet_pblock.c	(working copy)
@@ -502,7 +509,7 @@
     libnet_pblock_t *c;
     u_int32_t ip_offset = 0;
 
-    assert(p->type == LIBNET_PBLOCK_IPV4_H);
+    assert(p->type == LIBNET_PBLOCK_IPV4_H || p->type == LIBNET_PBLOCK_IPV6_H);
 
     for(c = p; c; c = c->prev)
         ip_offset += c->b_len;
