diff -urNp libnet-1.1.4.org/include/libnet.h.in libnet-1.1.4/include/libnet.h.in
--- libnet-1.1.4.org/include/libnet.h.in	2010-03-01 21:57:18.000000000 +0200
+++ libnet-1.1.4/include/libnet.h.in	2010-03-06 22:02:14.898423826 +0200
@@ -113,6 +113,13 @@ extern "C" {
 #include <errno.h>
 #include <stdarg.h>
 
+/* workaround symbol conflict (found in AIX 5.3) */
+#undef ip_len
+#undef ip_v
+#undef ip_hl
+#undef ip_tos
+#undef th_win
+
 #define LIBNET_VERSION  "@PACKAGE_VERSION@"
 
 #define @ENDIANESS@ 1
